### DISCUSSION TITLE
I have 3 questions that I need advice on

### DISCUSSION THREAD

**Discussion by rushmi0** [quarkusio/quarkus#43430] [Q&A]
### 1. Receiving HTTP Headers
I want to receive the client's HTTP headers to use in the `onOpen` function. Here's an example of the header:

```
GET http://localhost:6724
accept: application/lnwza007+json
```

### 2. Opening the website with the root URL
Currently, I need to use this URL to open the website:
```
http://localhost:6724/index.html
```

But I want to change it so that the web landing page can be opened using the root URL like this:
```
http://localhost:6724/
```

**I want to ues WebSocket and HTTP to work through the same endpoint**

```
GET http://localhost:6724
accept: application/lnwza007+json

###

WEBSOCKET ws://localhost:6724
Content-Type: application/json

["hsZEOtaDsENYkP5H-JIWp"]
```

### 3. Setting up CORS
I’ve set up CORS according to the documentation, but it's still not working:
https://cors-test.codehappy.dev

repo:
https://github.com/rushmi0/quarkus-websocket/blob/main/src/main/kotlin/org/demo/Gateway.kt

https://github.com/rushmi0/quarkus-websocket/blob/main/src%2Fmain%2Fkotlin%2Forg%2Fdemo%2FServerInfo.kt

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
cc @mkouba

**Comment by mkouba** [comment]
> # 1. Receiving HTTP Headers
> I want to receive the client's HTTP headers to use in the `onOpen` function. Here's an example of the header:

Just add the `io.quarkus.websockets.next.HandshakeRequest` param in your `@OnOpen` callback and then use one of the convenient methods to access the headers of the the initial handshake request.

> # 2. Opening the website with the root URL
> I want to ues WebSocket and HTTP to work through the same endpoint

This is not supported at the moment because we always handle a WebSocket as an HTTP Upgrade. I'm not sure if that's even feasible but I suppose that we would have to use the `HttpServer#webSocketHandler(Handler<ServerWebSocket>)` instead of  route per endpoint + `HttpServerRequest#toWebSocket()` @cescoffier

> ### 3. Setting up CORS
> I’ve set up CORS according to the documentation, but it's still not working: https://cors-test.codehappy.dev

What documentation do you mean?

**Reply by cescoffier** [reply]
> This is not supported at the moment because we always handle a WebSocket as an HTTP Upgrade. I'm not sure if that's even feasible but I suppose that we would have to use the HttpServer#webSocketHandler(Handler<ServerWebSocket>) instead of route per endpoint + HttpServerRequest#toWebSocket() @cescoffier

You can receive an HTTP request and upgrade it to a web socket or process it as a _plain_ HTTP method. However, I would not recommend using this, and use two different paths (for simplicity purpose).

**Reply by mkouba** [reply]
> You can receive an HTTP request and upgrade it to a web socket or process it as a plain HTTP method.

Hm, we could probably use the next route if the `Sec-WebSocket-Key` header is not set, right?

> However, I would not recommend using this, and use two different paths (for simplicity purpose).

:+1:

**Reply by mkouba** [reply]
FYI https://github.com/quarkusio/quarkus/pull/43439

**Reply by rushmi0** [reply]
> > # 1. Receiving HTTP Headers
> > I want to receive the client's HTTP headers to use in the `onOpen` function. Here's an example of the header:
> 
> Just add the `io.quarkus.websockets.next.HandshakeRequest` param in your `@OnOpen` callback and then use one of the convenient methods to access the headers of the the initial handshake request.
> 
> > # 2. Opening the website with the root URL
> > I want to ues WebSocket and HTTP to work through the same endpoint
> 
> This is not supported at the moment because we always handle a WebSocket as an HTTP Upgrade. I'm not sure if that's even feasible but I suppose that we would have to use the `HttpServer#webSocketHandler(Handler<ServerWebSocket>)` instead of route per endpoint + `HttpServerRequest#toWebSocket()` @cescoffier
> 
> > ### 3. Setting up CORS
> > I’ve set up CORS according to the documentation, but it's still not working: https://cors-test.codehappy.dev
> 
> What documentation do you mean?

@mkouba 
This document: https://quarkus.io/guides/security-cors

**Reply by mkouba** [reply]
> This document: https://quarkus.io/guides/security-cors

Ok, what exactly does not work? 

CC @sberyozkin

**Reply by rushmi0** [reply]
@mkouba 
![image](https://github.com/user-attachments/assets/3f937920-f48b-4bb1-9b10-9450427527d2)


![image](https://github.com/user-attachments/assets/54979417-98cb-45f3-8003-ea027ca84b3b)
